lapce-core/src/language.rs

```diff
@@ -3,930 +3,1810 @@ use std::{collections::HashSet, path::Path, str::FromStr};
+    // multi_line_comment_token: &'static str,
+    // /// All tokens that can be used for comments in language
+    // comment_tokens: &'static [&'static str],
+    comment: CommentProperties,
+
+    /// The indent unit.  
+    /// "  " for bash, "    " for rust, for example.
+    indent: &'static str,
+    /// Filenames that belong to this language  
+    /// `["Dockerfile"]` for Dockerfile, `[".editorconfig"]` for EditorConfig
+    files: &'static [&'static str],
+    /// File name extensions to determine the language.  
+    /// `["py"]` for python, `["rs"]` for rust, for example.
+    extensions: &'static [&'static str],
+    /// Tree-sitter properties
+    tree_sitter: Option<TreeSitterProperties>,
+}
+
+#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug, PartialOrd, Ord)]
+struct TreeSitterProperties {
+    /// This is the factory function defined in the tree-sitter crate that creates the language parser.  
+    /// For most languages, it is `tree_sitter_$crate::language`.
     language: fn() -> tree_sitter::Language,
     /// For most languages, it is `tree_sitter_$crate::HIGHLIGHT_QUERY`.
-    highlight: &'static str,
-    /// For most languages, it is `tree_sitter_$crate::INJECTION_QUERY`.
+    highlight: Option<&'static str>,
+    /// For most languages, it is `tree_sitter_$crate::INJECTION_QUERY`.  
     /// Though, not all languages have injections.
     injection: Option<&'static str>,
-    /// The comment token.  "#" for python, "//" for rust for example.
-    comment: &'static str,
-    /// The indent unit.  "  " for javascript, "    " for rust, for example.
-    indent: &'static str,
     /// TODO: someone more knowledgeable please describe what the two lists are.
     /// Anyway, the second element of the tuple is a "ignore list". See
     /// `walk_tree`. If unsure, use `DEFAULT_CODE_LENS_LIST` and
     /// `DEFAULT_CODE_LENS_IGNORE_LIST`.
     code_lens: (&'static [&'static str], &'static [&'static str]),
     /// the tree sitter tag names that can be put in sticky headers
     sticky_headers: &'static [&'static str],
-    /// File name extensions to determine the language.  `["py"]` for python,
-    /// `["rs"]` for rust, for example.
-    extensions: &'static [&'static str],
 }
 
-// NOTE: Keep the enum variants "fieldless" so they can cast to usize as array
-// indices into the LANGUAGES array.  See method `LapceLanguage::properties`.
-//
-// Do not assign values to the variants because the number of variants and
-// number of elements in the LANGUAGES array change as different features
-// selected by the cargo build command.
-#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug, Display, EnumString)]
+#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug, PartialOrd, Ord, Default)]
+struct CommentProperties {
+    single_line_start: &'static str,
+    single_line_end: &'static str,
+
+    multi_line_start: &'static str,
+    multi_line_end: &'static str,
+    multi_line_prefix: &'static str,
+}
```

Comment:
A lot of the languages in this file only have a `single_line_start`. Making this an enum w/ "simple"/"conplex" variants would cut down on that boilerplate:
```suggestion
enum CommentProperties {
    /// A simple comment style where all comments are
    /// single line comments beginning with the same token.
    Simple {
            single_line_start: &'static str,
    },
    
    /// A more complicated comment style where comments
    /// may not all begin and end with the same token.
    Complex {
        single_line_start: &'static str,
        single_line_end: &'static str,

        multi_line_start: &'static str,
        multi_line_end: &'static str,
        multi_line_prefix: &'static str,
    }
}
```

And then a lot of the languages can be simplified to eg `CommentProperties::Simple("//")` or `CommentProperties::Simple("#")`